### ISSUE TITLE
Feature: add detection for system network changes and auto-reload config

### ISSUE THREAD

**Issue by bradh352** [c-ares/c-ares#613]
This would likely depend on #611. We would want to automatically call ares_reinit() on any detected system change.

On Linux, we'd probably need to listen on DBUS for NetworkManager changes which would probably mean linking to either libdbus or system'd dbus implementation (systemd's honestly is a lot nicer, surprisingly).  We could alternatively monitor NETLINK_ROUTE, but that's considered more legacy but wouldn't require depending on other libraries.

On Windows, we'd need to register events with NotifyAddrChange() and NotifyRouteChange().

On MacOS, iOS, Android, etc we'd need to do more research and see what platforms we want to support.  Application developers can of course do the same thing themselves and just call ares_reinit() themselves if they detect a change.

**Comment by bagder** [comment]
Can all of those things still work when c-ares is only invoked once in a while from an application? How will c-ares know that the network changed between one invoke now and the previous one 30 minutes ago? c-ares has no background process that can monitor the situation.

**Comment by bradh352** [comment]
Yeah, if #611 is implemented (the eventloop thread), it would have a thread constantly running for any `ares_channel_t` (optionally, of course ... but it would be a requirement to use that feature in order to get this one).

**Comment by bradh352** [comment]
implemented in PR #759
